### ISSUE TITLE
False alarm for TRIO108 in context managers

### ISSUE THREAD

**Issue by Zac-HD** [Zac-HD/flake8-trio#39]
```python
from contextlib import asynccontextmanager
import trio


@asynccontextmanager
async def in_background(*background_tasks):
    async with trio.open_nursery() as nursery:
        for bgt in background_tasks:
            nursery.start_soon(bgt)
        yield
```
doesn't emit TRIO101 because we know this is (probably) a context manager, but we don't have the same filter for 108:
```
$ flake8 example.py 
example.py:6:1: TRIO108: exit from async iterable with no guaranteed checkpoint since yield on line 10
example.py:10:9: TRIO108: yield from async iterable with no guaranteed checkpoint since function definition on line 6
```

**Comment by jakkdl** [comment]
Ah okay, it's currently very conservative and treats decorators the same as `async with` - i.e. guaranteed checkpoint on at least one of entry and exit. I remember you responding this somewhere when I asked, but can't find that comment now.
So it's not about not having the filter, and 107 isn't triggered since we know it'll checkpoint at least once. But it's likely overly conservative though, so I'll switch 108 to seeing function decorator context managers as checkpointing both on entry and exit.

**Comment by Zac-HD** [comment]
Yeah, I think the general principle is "whichever interpretation is less prone to false alarms"; if we don't know which then assuming it's fine seems like a better user experience.

**Comment by jakkdl** [comment]
I could also treat `async with` as checkpointing on both entry and exit in that case, such that
```python
async def foo():
  async with ...:
    yield
```
raises no errors.

**Comment by Zac-HD** [comment]
I think that would reduce false alarms,  so let's do it 🙂

**Comment by jakkdl** [comment]
It hurts a bit to remove that beautiful logic, it could plausibly be put behind a setting. But I don't think the case is strong enough to bother with anything like that and in 99% of cases it's probably false alarms. RIP `self.yield_count`
